c/meterpreter/source/metsrv/pool_party.c

```diff
@@ -0,0 +1,299 @@
+			if (ntdll->pZwSetIoCompletion != NULL) {
+				if (poolLifeguard != NULL) {
+					poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_DIRECT_INSERTION].isSystemSupported = TRUE;
+				}
+			}
+			dprintf("[INJECT][inject_via_poolparty][ntdll_init] ZwSetIoCompletion: %p", ntdll->pZwSetIoCompletion);
+
+			//ntdll->pZwAssociateWaitCompletionPacket = (NTSTATUS(NTAPI*)(HANDLE, HANDLE, HANDLE, PVOID, PVOID, NTSTATUS, ULONG_PTR, PBOOLEAN))GetProcAddress(hNtDll, "ZwAssociateWaitCompletionPacket");
+			//if (ntdll->pZwAssociateWaitCompletionPacket != NULL) {
+			//	if (poolLifeguard != NULL) {
+			//		poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_WAIT_INSERTION].isSystemSupported = TRUE;
+			//	}
+			//}
+			//dprintf("[INJECT][inject_via_poolparty][ntdll_init] ZwAssociateWaitCompletionPacket: %p", ntdll->pZwAssociateWaitCompletionPacket);
+
+			//ntdll->pNtQueryInformationWorkerFactory = (NTSTATUS(NTAPI*)(HANDLE, _WORKERFACTORYINFOCLASS, PVOID, ULONG, PULONG))GetProcAddress(hNtDll, "NtQueryInformationWorkerFactory"); // WIN 7
+			//dprintf("[INJECT][inject_via_poolparty][ntdll_init] NtQueryInformationWorkerFactory: %p", ntdll->pNtQueryInformationWorkerFactory);
+
+			//ntdll->pNtSetInformationWorkerFactory = (NTSTATUS(NTAPI*)(HANDLE, _WORKERFACTORYINFOCLASS, PVOID, ULONG))GetProcAddress(hNtDll, "NtSetInformationWorkerFactory"); // WIN7
+			//dprintf("[INJECT][inject_via_poolparty][ntdll_init] NtSetInformationWorkerFactory: %p", ntdll->pNtSetInformationWorkerFactory);
+
+			//if (ntdll->pNtQueryInformationWorkerFactory != NULL && ntdll->pNtSetInformationWorkerFactory != NULL) {
+			//	if (poolLifeguard != NULL) {
+			//		poolLifeguard->variants[POOLPARTY_TECHNIQUE_WORKER_FACTORY_OVERWRITE].isSystemSupported = TRUE;
+			//	}
+			//}
+		}
+	}
+	return ntdll;
+}
+
+POOLPARTY_INJECTOR* GetOrInitPoolParty(DWORD dwSourceArch, DWORD dwDestinationArch) {
+	BOOL bError = FALSE;
+	HANDLE hHeap = GetProcessHeap();
+	bError = (hHeap == NULL);
+	BOOL isWow64;
+	IsWow64Process(GetCurrentProcess(), &isWow64);
+
+	if (poolLifeguard != NULL) {
+		return poolLifeguard;
+	}
+
+	if (!bError) {
+		poolLifeguard = (POOLPARTY_INJECTOR*)HeapAlloc(hHeap, HEAP_ZERO_MEMORY, sizeof(POOLPARTY_INJECTOR));
+		bError = poolLifeguard == NULL;
+		
+		if(!bError) {
+			poolLifeguard->init = FALSE;
+			if (ntdll == NULL) {
+				bError = GetOrInitNtDll() == NULL;
+			}
+		}
+
+		if (!bError) {
+			if (dwSourceArch == PROCESS_ARCH_X64) {
+				if (dwDestinationArch == PROCESS_ARCH_X64) {
+					// poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_WAIT_INSERTION].isInjectionSupported = poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_WAIT_INSERTION].isSystemSupported;
+				}
+				poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_DIRECT_INSERTION].isInjectionSupported = poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_DIRECT_INSERTION].isSystemSupported;
+			}
+		}
+
+		if (!bError) {
```

Comment:
+1 on this.  We do a LOT of unholy things in meterpreter code to create a run-once loop to aid in error handling.  I think they're all kind of an abomination, but the `while(TRUE)` is the least awful because it telegraphs that is being abused for utility at the outset.  Above you use a `do...while` and they're used a lot, but I've hated them especially because when I see a `do` I expect a loop and don't discover that it is being abused until later.  When I see `while(TRUE)` I accept at the very beginning that some sort of shenanigans are afoot.

Reply:
I think I didn't see a `while (TRUE)` used as a `try-catch` statement but I saw and used (even if I don't really like it) the `do .. while(0)`. Should I use a `while(TRUE)` even if would be the non conventional way for this C code or keep the `do ... while` ? I **personally** prefer the `do...while` since I use the `while(TRUE)` as a real loop and not to guard expception, (for example like in the `GetRemoteHandles`.